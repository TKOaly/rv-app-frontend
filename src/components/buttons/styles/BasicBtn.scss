@import "vars";

/*
  Basic buttons
  */

.basic {
  color: $basic-button-bgcolor;
  -webkit-transition: background-color 0.1s; /* Safari */
  transition: background-color 0.1s;
}

.basic-hover:hover {
  color: $basic-button-textcolor;
  background-color: $basic-button-bgcolor;
  border-color: $basic-button-bgcolor;
  cursor: pointer;
}

.basic-fill {
  color: $basic-button-textcolor;
  background-color: $basic-button-bgcolor;
  border-color: $basic-button-bgcolor;
  -webkit-transition: background-color 0.1s; /* Safari */
  transition: background-color 0.1s;
}

.basic-fill-hover:hover {
  color: $basic-button-bgcolor;
  background-color: transparent;
  border-color: $basic-button-bgcolor;
  cursor: pointer;
}
