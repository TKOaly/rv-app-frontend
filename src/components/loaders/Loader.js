import './styles/Loader.css';
import React from 'react';

const Loader = () => <div className="loader" />;

export default Loader;
