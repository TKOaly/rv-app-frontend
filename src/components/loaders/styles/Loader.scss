@import "vars";

.loader {
  border: 2px solid transparent;
  border-top: $loader-width solid $loader-color;
  border-radius: 50%;
  width: 20px;
  height: 20px;
  animation: spin 0.5s linear infinite;
  margin-left: auto;
  margin-right: auto;
}
@keyframes spin {
  
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}
