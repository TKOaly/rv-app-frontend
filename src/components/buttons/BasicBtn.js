import './styles/BasicBtn.css';
import './styles/Button.css';
import Loader from '../loaders/Loader';
import PropTypes from 'prop-types';
import React from 'react';

/**
 * Basic button.
 *
 * Suitable for UI actions, that do not change the current page or execute a critical action, such as submitting a form.
 */
const BasicBtn = ({ onClick, children, fill, hover, loader, ...props }) => {
    let className = 'btn';
    fill ? (className += ' basic-fill') : (className += ' basic');
    hover && (fill ? (className += ' basic-fill-hover') : (className += ' basic-hover'));

    return (
        <button {...props} onClick={onClick} className={className}>
            {!loader ? <span className="btnContent">{children}</span> : <Loader />}
        </button>
    );
};

BasicBtn.propTypes = {
    /** Click handler. */
    onClick: PropTypes.func,
    /** Button fill. If set to `true`, the button's colors will "invert". */
    fill: PropTypes.bool,
    /** Hover effect. If set to `true`, the button will have a hover effect. */
    hover: PropTypes.bool,
    /** Loader effect. If set to `true`, will show a CSS loader instead of button text. */
    loader: PropTypes.bool
};

BasicBtn.defaultProps = {
    onClick: null,
    fill: false,
    hover: false,
    loader: false
};

export default BasicBtn;
