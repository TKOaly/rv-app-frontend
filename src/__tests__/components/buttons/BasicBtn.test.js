import { mount, shallow } from 'enzyme';
import BasicBtn from '../../../components/buttons/BasicBtn';
import Loader from '../../../components/loaders/Loader';
import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

describe('<BasicBtn />', () => {
    it('should render text correctly', () => {
        const basicBtn = shallow(<BasicBtn>Hello world</BasicBtn>);
        const contentDiv = basicBtn.find('.btnContent');
        expect(contentDiv.text()).toContain('Hello world');
    });

    it('should match snapshot', () => {
        const cmpnt = renderer.create(<BasicBtn>Hello world</BasicBtn>).toJSON();
        expect(cmpnt).toMatchSnapshot();
    });

    it('should show loader correctly', () => {
        const dangerBtnComponent = shallow(<BasicBtn loader />);
        expect(dangerBtnComponent.contains(<Loader />)).toBe(true);
    });

    it('should fire an event when button is clicked', () => {
        const mockHandler = sinon.spy();

        const basicBtnComponent = mount(<BasicBtn onClick={mockHandler} />);

        expect(mockHandler.notCalled).toBe(true);
        basicBtnComponent.simulate('click');
        expect(mockHandler.calledOnce).toBe(true);
    });
});
