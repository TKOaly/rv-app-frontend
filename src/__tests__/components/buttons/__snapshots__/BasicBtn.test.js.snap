// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BasicBtn /> should match snapshot 1`] = `
<button
  className="btn basic"
  onClick={null}
>
  <span
    className="btnContent"
  >
    Hello world
  </span>
</button>
`;
